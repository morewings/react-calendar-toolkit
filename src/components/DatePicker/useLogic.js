import {useCallback} from 'react';
import {
  useDatePickerActions,
  useDatePickerContext,
  useIncrementPrecision,
} from 'features/datepicker';

const useLogic = (onDateSet, minPrecision) => {
  const {
    state: {precision},
  } = useDatePickerContext();
  const {setDate, setVisibility} = useDatePickerActions();
  const incrementPrecision = useIncrementPrecision(minPrecision);
  const handleDateSet = useCallback(
    date => {
      if (precision === minPrecision) {
        onDateSet(date);
        setDate(date);
      } else {
        setVisibility(date);
        incrementPrecision();
      }
    },
    [
      incrementPrecision,
      minPrecision,
      onDateSet,
      precision,
      setDate,
      setVisibility,
    ]
  );
  return {handleDateSet};
};

export default useLogic;
