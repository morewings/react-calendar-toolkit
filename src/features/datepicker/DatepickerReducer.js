import {SET_DATE, SET_TODAY, SET_PRECISION} from './actionTypes';

const initialState = {
  selectedTimestamp: 0,
  todayTimestamp: 0,
  precision: 'day',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_DATE: {
      const {selectedTimestamp, precision} = action.payload;
      return {
        ...state,
        selectedTimestamp,
        precision,
      };
    }
    case SET_TODAY: {
      const {todayTimestamp} = action.payload;
      return {
        ...state,
        todayTimestamp,
      };
    }
    case SET_PRECISION: {
      return {
        ...state,
        precision: action.payload,
      };
    }
    default:
      return state;
  }
};
