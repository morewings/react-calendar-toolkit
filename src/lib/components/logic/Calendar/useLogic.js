import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsSameMonth,
  checkIsWeekend,
  checkIsWithinInterval,
  floorMonth,
  matchDatesWithPrecision,
} from 'lib/utils/dateUtils';

const useLogic = ({
  disableDate,
  endDate,
  precision,
  startDate,
  highlightDate,
  onDateSet,
  highlightWeekends,
  visibleTimestamp,
  selectedTimestamp,
  todayTimestamp,
}) => {
  const getIsDisabled = useCallback(
    date => {
      if (precision === 'year') return false; // we are not disabling years, just rendering the range
      if (precision === 'month') {
        return !checkIsWithinInterval(
          {start: floorMonth(startDate), end: ceilMonth(endDate)},
          date
        );
      }
      return (
        disableDate({date, isWeekend: checkIsWeekend(date), precision}) ||
        !checkIsWithinInterval({start: startDate, end: endDate}, date)
      );
    },
    [disableDate, endDate, precision, startDate]
  );

  const getIsHighlighted = useCallback(
    date => highlightDate({date, isWeekend: checkIsWeekend(date), precision}),
    [highlightDate, precision]
  );

  const handleDateSet = useCallback(
    date => {
      const isDisabled = getIsDisabled(date);
      !isDisabled && onDateSet(date);
    },
    [getIsDisabled, onDateSet]
  );

  const getIsWeekend = date =>
    precision === 'day' && highlightWeekends && checkIsWeekend(date);

  const getBelongsToSameMonth = date =>
    precision === 'day' && checkIsSameMonth(date, visibleTimestamp);

  const getIsSelected = date =>
    matchDatesWithPrecision(precision, date, selectedTimestamp);

  const getIsCurrent = date =>
    matchDatesWithPrecision(precision, date, todayTimestamp);

  return {
    getIsDisabled,
    getIsHighlighted,
    handleDateSet,
    getIsWeekend,
    getBelongsToSameMonth,
    getIsSelected,
    getIsCurrent,
  };
};

export default useLogic;
