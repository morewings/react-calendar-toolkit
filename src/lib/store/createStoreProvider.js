import React, {useReducer} from 'react';

const createStoreProvider = ({initialState, reducer, context}) => ({
  // eslint-disable-next-line react/prop-types
  children,
}) => {
  const {Provider} = context;
  const [state, dispatch] = useReducer(reducer, initialState);
  return <Provider value={{state, dispatch}}>{children}</Provider>;
};

export default createStoreProvider;
