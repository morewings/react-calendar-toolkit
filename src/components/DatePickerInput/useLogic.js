import {useCallback} from 'react';
import {useModalActions} from 'features/modal';
import {convertToDate} from 'utils/dateUtils';
import {useFormatDate} from 'features/locale';
import {useDatePickerContext} from 'features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatePicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatePickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatePicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatePicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
