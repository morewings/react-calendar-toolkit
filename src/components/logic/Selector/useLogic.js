import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  decrementMonth,
  floorMonth,
  incrementMonth,
} from 'utils/dateUtils';
import {useDatePickerActions} from 'features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatePickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = incrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = decrementMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
