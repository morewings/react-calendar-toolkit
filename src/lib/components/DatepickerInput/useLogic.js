import {useCallback} from 'react';
import {useModalActions} from 'lib/features/modal';
import {convertToDate} from 'lib/utils/dateUtils';
import {useFormatDate} from 'lib/features/locale';
import {useDatepickerContext} from 'lib/features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatepicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatepickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatepicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatepicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
