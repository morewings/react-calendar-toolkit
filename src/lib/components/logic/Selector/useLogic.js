import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  subMonth,
  floorMonth,
  addMonth,
} from 'lib/utils/dateUtils';
import {useDatepickerActions} from 'lib/features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatepickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = addMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = subMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
