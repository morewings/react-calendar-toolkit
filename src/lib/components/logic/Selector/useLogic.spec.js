import React from 'react';
import {renderHook, act} from '@testing-library/react-hooks';
import createStoreProvider from 'lib/store';
import DatepickerContext from 'lib/features/datepicker/context';
import DatepickerReducer from 'lib/features/datepicker/DatepickerReducer';
import useLogic from './useLogic';

const mockReducer = jest.fn((...args) => {
  DatepickerReducer(...args);
});

const visibleTimestamp = new Date(2020, 0, 12).getTime(); // // 12.01.2020
const startDate = new Date(2019, 0, 12); // // 12.01.2020
const endDate = new Date(2021, 0, 12); // // 12.01.2020

const createStoreProviderWithState = state =>
  createStoreProvider({
    initialState: state,
    reducer: mockReducer,
    context: DatepickerContext,
  });

describe('Selector > useLogic ', () => {
  beforeEach(() => {
    mockReducer.mockClear();
  });

  it('returns functions', () => {
    const Provider = createStoreProviderWithState({});
    const {result} = renderHook(
      () => useLogic({visibleTimestamp, endDate, startDate}),
      {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      }
    );
    expect(result.current).toMatchSnapshot();
  });

  describe('onSetPrecision', () => {
    it('sets precision', () => {
      const Provider = createStoreProviderWithState({
        precision: 'day',
      });
      const {result} = renderHook(
        () => useLogic({visibleTimestamp, endDate, startDate}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => {
        result.current.onSetPrecision('month');
      });
      expect(mockReducer).toHaveBeenCalledTimes(1);
      expect(mockReducer.mock.calls[0][1]).toMatchSnapshot(); // SET_PRECISION
    });
  });

  describe('onIncrementMonth', () => {
    it('sets next visible date by adding 1 to current, if next date is within provided range', () => {
      const Provider = createStoreProviderWithState({});
      const {result} = renderHook(
        () => useLogic({visibleTimestamp, endDate, startDate}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => {
        result.current.onIncrementMonth();
      });
      expect(mockReducer).toHaveBeenCalledTimes(1);
      expect(mockReducer.mock.calls[0][1]).toMatchSnapshot(); // SET_VISIBILITY
    });

    it('does not set next visible date, if next date is out of provided range', () => {
      const Provider = createStoreProviderWithState({});
      const {result} = renderHook(
        () =>
          useLogic({visibleTimestamp, endDate: visibleTimestamp, startDate}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => {
        result.current.onIncrementMonth();
      });
      expect(mockReducer).not.toHaveBeenCalled();
    });
  });

  describe('onDecrementMonth', () => {
    it('sets next visible date by subtracting 1 from current, if next date is within provided range', () => {
      const Provider = createStoreProviderWithState({});
      const {result} = renderHook(
        () => useLogic({visibleTimestamp, endDate, startDate}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => {
        result.current.onDecrementMonth();
      });
      expect(mockReducer).toHaveBeenCalledTimes(1);
      expect(mockReducer.mock.calls[0][1]).toMatchSnapshot(); // SET_VISIBILITY
    });

    it('does not set next visible date, if next date is out of provided range', () => {
      const Provider = createStoreProviderWithState({});
      const {result} = renderHook(
        () =>
          useLogic({visibleTimestamp, endDate, startDate: visibleTimestamp}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => {
        result.current.onDecrementMonth();
      });
      expect(mockReducer).not.toHaveBeenCalled();
    });
  });
});
