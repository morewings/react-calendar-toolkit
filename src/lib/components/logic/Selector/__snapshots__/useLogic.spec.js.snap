// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector > useLogic  onDecrementMonth sets next visible date by subtracting 1 from current, if next date is within provided range 1`] = `
Object {
  "payload": Object {
    "visibleTimestamp": 1576108800000,
  },
  "type": "SET_VISIBILITY",
}
`;

exports[`Selector > useLogic  onIncrementMonth sets next visible date by adding 1 to current, if next date is within provided range 1`] = `
Object {
  "payload": Object {
    "visibleTimestamp": 1581465600000,
  },
  "type": "SET_VISIBILITY",
}
`;

exports[`Selector > useLogic  onSetPrecision sets precision 1`] = `
Object {
  "payload": "month",
  "type": "SET_PRECISION",
}
`;

exports[`Selector > useLogic  returns functions 1`] = `
Object {
  "onDecrementMonth": [Function],
  "onIncrementMonth": [Function],
  "onSetPrecision": [Function],
}
`;
